app/src/main/java/com/keylesspalace/tusky/util/HttpHeaderLink.kt

```diff
@@ -0,0 +1,150 @@
+                result.end = foundIndex
+                result.value = line.substring(start, foundIndex)
+                result
+            } else {
+                var quoteEnd = findEndOfQuotedString(line, foundIndex + 1)
+                if (quoteEnd == -1) {
+                    quoteEnd = line.length
+                }
+                result.end = quoteEnd
+                result.value = line.substring(foundIndex + 1, quoteEnd)
+                result
+            }
+        }
+
+        private fun parseParameters(line: String, start: Int, link: HttpHeaderLink): Int {
+            var i = start
+            while (i < line.length) {
+                val foundIndex = line.indexOfAny(charArrayOf('=', ','), i, false)
+                if (foundIndex == -1) {
+                    return -1
+                } else if (line[foundIndex] == ',') {
+                    return foundIndex
+                }
+                val parameter = Parameter()
+                parameter.name =
+                    line.substring(line.indexOf(';', i) + 1, foundIndex).trim()
+                link.parameters.add(parameter)
+                val result = parseValue(line, foundIndex)
+                parameter.value = result.value
+                i = if (result.end == -1) {
+                    return -1
+                } else {
+                    result.end
+                }
+            }
+            return -1
+        }
+
+        /**
+         * @param line the entire link header, not including the initial "Link:"
+         * @return all links found in the header
+         */
+        fun parse(line: String?): List<HttpHeaderLink> {
+            val linkList: MutableList<HttpHeaderLink> = ArrayList()
+            line ?: return linkList
+
+            var i = 0
+            while (i < line.length) {
+                val uriEnd = line.indexOf('>', i)
+                val uri = line.substring(line.indexOf('<', i) + 1, uriEnd)
+                val link = HttpHeaderLink(uri)
+                linkList.add(link)
+                val parseEnd = parseParameters(line, uriEnd, link)
+                i = if (parseEnd == -1) {
+                    break
+                } else {
+                    parseEnd
+                }
+                i++
+            }
+
+            return linkList
+        }
+
+        /**
+         * @param links intended to be those returned by parse()
+         * @param relationType of the parameter "rel", commonly "next" or "prev"
+         * @return the link matching the given relation type
+         */
+        fun findByRelationType(
+            links: List<HttpHeaderLink>,
+            relationType: String
+        ): HttpHeaderLink? {
+            for (link in links) {
+                for (parameter in link.parameters) {
+                    if (parameter.name == "rel" && parameter.value == relationType) {
+                        return link
+                    }
+                }
+            }
+            return null
```

Comment:
You can write this in nice idiomatic Kotlin like so:

```suggestion
         return links.find { link ->
                link.parameters.any { parameter ->
                    parameter.name == "rel" && parameter.value == relationType
                }
            }
```